File: backend/app/schemas/personnel.py
"""Personnel management schemas."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class OrgNodePayload(BaseModel):
    id: str | None = None
    parent_id: str | None = None
    title: str = Field(min_length=1, max_length=200)
    role_title: str | None = None
    person_name: str | None = None
    phone: str | None = None
    email: str | None = None
    pos_x: float = 0
    pos_y: float = 0
    sort_order: int = 0


class OrgLinkPayload(BaseModel):
    id: str | None = None
    source_node_id: str
    target_node_id: str


class OrgChartCreate(BaseModel):
    project_no: str = Field(min_length=1, max_length=100)
    name: str = Field(min_length=1, max_length=200)


class OrgChartUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    project_no: str | None = Field(default=None, min_length=1, max_length=100)
    nodes: list[OrgNodePayload] | None = None
    links: list[OrgLinkPayload] | None = None
    canvas_json: dict | list | None = None


class OrgNodeResponse(BaseModel):
    id: str
    chart_id: str
    parent_id: str | None = None
    title: str
    role_title: str | None = None
    person_name: str | None = None
    phone: str | None = None
    email: str | None = None
    pos_x: float
    pos_y: float
    sort_order: int


class OrgLinkResponse(BaseModel):
    id: str
    chart_id: str
    source_node_id: str
    target_node_id: str


class OrgChartBrief(BaseModel):
    id: str
    project_no: str
    name: str
    node_count: int = 0
    created_at: datetime
    updated_at: datetime


class OrgChartResponse(BaseModel):
    id: str
    project_no: str
    name: str
    canvas_json: dict | list | None = None
    nodes: list[OrgNodeResponse] = Field(default_factory=list)
    links: list[OrgLinkResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class InternalCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    role_title: str = Field(default="", max_length=100)
    phone: str | None = None
    email: str | None = None
    company: str | None = None
    project_no: str | None = None
    org_node_id: str | None = None
    notes: str | None = None


class InternalUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    role_title: str | None = None
    phone: str | None = None
    email: str | None = None
    company: str | None = None
    project_no: str | None = None
    org_node_id: str | None = None
    notes: str | None = None


class InternalResponse(BaseModel):
    id: str
    name: str
    role_title: str
    phone: str | None = None
    email: str | None = None
    company: str | None = None
    project_no: str | None = None
    org_node_id: str | None = None
    notes: str | None = None
    created_at: datetime
    updated_at: datetime


class SupplierProductPayload(BaseModel):
    device_model_id: str | None = None
    device_name: str | None = None
    device_model_name: str | None = None


class SupplierCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    role_title: str = Field(default="", max_length=100)
    phone: str | None = None
    email: str | None = None
    wechat: str | None = None
    manufacturer_id: str
    notes: str | None = None
    contract_ids: list[str] = Field(default_factory=list)
    products: list[SupplierProductPayload] = Field(default_factory=list)


class SupplierUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    role_title: str | None = None
    phone: str | None = None
    email: str | None = None
    wechat: str | None = None
    manufacturer_id: str | None = None
    notes: str | None = None
    contract_ids: list[str] | None = None
    products: list[SupplierProductPayload] | None = None


class SupplierProductResponse(BaseModel):
    id: str
    device_model_id: str | None = None
    device_name: str | None = None
    device_model_name: str | None = None


class SupplierResponse(BaseModel):
    id: str
    name: str
    role_title: str
    phone: str | None = None
    email: str | None = None
    wechat: str | None = None
    manufacturer_id: str
    manufacturer_name: str | None = None
    notes: str | None = None
    contract_ids: list[str] = Field(default_factory=list)
    contract_nos: list[str] = Field(default_factory=list)
    products: list[SupplierProductResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

File: frontend/src/api/personnel.ts
import api, { unwrap } from '@/api'
import type { ApiResponse } from '@/types/api'

export interface OrgChartBrief {
  id: string
  project_no: string
  name: string
  node_count: number
  created_at: string
  updated_at: string
}

export interface OrgNode {
  id: string
  chart_id: string
  parent_id: string | null
  title: string
  role_title: string | null
  person_name: string | null
  phone: string | null
  email: string | null
  pos_x: number
  pos_y: number
  sort_order: number
}

export interface OrgLink {
  id: string
  chart_id: string
  source_node_id: string
  target_node_id: string
}

export interface OrgChart {
  id: string
  project_no: string
  name: string
  canvas_json: Record<string, unknown> | unknown[] | null
  nodes: OrgNode[]
  links: OrgLink[]
  created_at: string
  updated_at: string
}

export interface InternalContact {
  id: string
  name: string
  role_title: string
  phone: string | null
  email: string | null
  company: string | null
  project_no: string | null
  org_node_id: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface SupplierProduct {
  id?: string
  device_model_id?: string | null
  device_name?: string | null
  device_model_name?: string | null
}

export interface SupplierContact {
  id: string
  name: string
  role_title: string
  phone: string | null
  email: string | null
  wechat: string | null
  manufacturer_id: string
  manufacturer_name: string | null
  notes: string | null
  contract_ids: string[]
  contract_nos: string[]
  products: SupplierProduct[]
  created_at: string
  updated_at: string
}

export async function listOrgCharts(projectNo?: string) {
  const response = await api.get<ApiResponse<OrgChartBrief[]>>('/personnel/org-charts', {
    params: projectNo ? { project_no: projectNo } : undefined,
  })
  return unwrap(response)
}

export async function getOrgChart(id: string) {
  const response = await api.get<ApiResponse<OrgChart>>(`/personnel/org-charts/${id}`)
  return unwrap(response)
}

export async function createOrgChart(payload: { project_no: string; name: string }) {
  const response = await api.post<ApiResponse<OrgChart>>('/personnel/org-charts', payload)
  return unwrap(response)
}

export async function updateOrgChart(
  id: string,
  payload: {
    name?: string
    project_no?: string
    nodes?: Array<{
      id?: string
      parent_id?: string | null
      title: string
      role_title?: string | null
      person_name?: string | null
      phone?: string | null
      email?: string | null
      pos_x?: number
      pos_y?: number
      sort_order?: number
    }>
    links?: Array<{ id?: string; source_node_id: string; target_node_id: string }>
    canvas_json?: Record<string, unknown> | null
  },
) {
  const response = await api.put<ApiResponse<OrgChart>>(`/personnel/org-charts/${id}`, payload)
  return unwrap(response)
}

export async function deleteOrgChart(id: string) {
  await api.delete(`/personnel/org-charts/${id}`)
}

export async function listInternals(params: Record<string, unknown> = {}) {
  const response = await api.get('/personnel/internals', { params })
  return response.data.data as {
    items: InternalContact[]
    pagination: { page: number; page_size: number; total: number; pages: number }
  }
}

export async function createInternal(payload: Partial<InternalContact> & { name: string }) {
  const response = await api.post<ApiResponse<InternalContact>>('/personnel/internals', payload)
  return unwrap(response)
}

export async function updateInternal(id: string, payload: Partial<InternalContact>) {
  const response = await api.put<ApiResponse<InternalContact>>(`/personnel/internals/${id}`, payload)
  return unwrap(response)
}

export async function deleteInternal(id: string) {
  await api.delete(`/personnel/internals/${id}`)
}

export async function listSuppliers(params: Record<string, unknown> = {}) {
  const response = await api.get('/personnel/suppliers', { params })
  return response.data.data as {
    items: SupplierContact[]
    pagination: { page: number; page_size: number; total: number; pages: number }
  }
}

export async function createSupplier(payload: {
  name: string
  role_title?: string
  phone?: string | null
  email?: string | null
  wechat?: string | null
  manufacturer_id: string
  notes?: string | null
  contract_ids?: string[]
  products?: SupplierProduct[]
}) {
  const response = await api.post<ApiResponse<SupplierContact>>('/personnel/suppliers', payload)
  return unwrap(response)
}

export async function updateSupplier(
  id: string,
  payload: {
    name?: string
    role_title?: string
    phone?: string | null
    email?: string | null
    wechat?: string | null
    manufacturer_id?: string
    notes?: string | null
    contract_ids?: string[]
    products?: SupplierProduct[]
  },
) {
  const response = await api.put<ApiResponse<SupplierContact>>(`/personnel/suppliers/${id}`, payload)
  return unwrap(response)
}

export async function deleteSupplier(id: string) {
  await api.delete(`/personnel/suppliers/${id}`)
}
